### PR TITLE
feat: Log variable scaling in mlflow

### PR DESCRIPTION
**Pull request by jakob-schloer**
## Description

The variable scaling of the loss function is not trivial anymore since many scalers can be multiplied. To keep track of that, the final variable scaling is logged to mlfow hyperparameters (see image) 

image

## Type of Change

-   [x] New feature (non-breaking change which adds functionality)

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [ ] I have added tests that prove my fix is effective or that my feature works
-   [ ] I ran the [complete Pytest test](https://anemoi.readthedocs.io/projects/training/en/latest/dev/testing.html) suite locally, and they pass
-   [x] I have tested the changes on a single GPU
-   [x] I have tested the changes on multiple GPUs / multi-node setups
-   [ ] I have run the Benchmark Profiler against the old version of the code

### Documentation

-   [x] My code follows the style guidelines of this project
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by sahahner** [comment]
Hi Jakob, I think this is a very useful PR. Thank you very much for opening it. Would it be feasible as well to list the variable scaling in a normalised way? I am wondering whether it would be helpful to divide the per-variable scalers by the sum of all scalers. On the other hand, this might not be the best solution when adding new variables. Any thoughts?

**Comment by jakob-schloer** [comment]
> Would it be feasible as well to list the variable scaling in a normalised way? I am wondering whether it would be helpful to divide the per-variable scalers by the sum of all scalers.

I agree that it would be nice to have the percentage scaling, but this might also be a bit misleading because the variables are not all normalized exactly to std=1 (depending on the normalization set).

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 1]
```diff
@@ -525,3 +525,10 @@ def configure_optimizers(self) -> tuple[list[torch.optim.Optimizer], list[dict]]
         )
 
         return [optimizer], [{"scheduler": scheduler, "interval": "step"}]
+
+    def setup(self, stage: str) -> None:
+        """Lightning hook that is called after model is initialized but before training starts."""
+        # Log variable scaling on rank 0
+        # The two ifs should be separate, but are combined due to pre-commit hook
+        if stage == "fit" and self.trainer.is_global_zero:
+            self.logger.log_hyperparams({"variable_scaling": self._scaling_values})
```
Comment (jakob-schloer) [inline training/src/anemoi/training/train/forecaster/forecaster.py]: Any opinions on the keyword? Maybe "loss_scaling_variable" is better.

Comment (sahahner) [reply training/src/anemoi/training/train/forecaster/forecaster.py]: Can we also add a resulting or final to indicate it is not necessarily set by the user? For example:
`final_variable_loss_scaling.XXX`

**Comment by anaprietonem** [comment]
Hey @jakob-schloer thanks for the PR! could you link an example mlflow run to have a better look? Also just to flag server performance in terms of rendering the runs can be impacted by the number of hparams we log so we need to be mindful there. In your screenshot we are talking about ~100 variables logged ?

**Comment by jakob-schloer** [comment]
> Hey @jakob-schloer thanks for the PR! could you link an example mlflow run to have a better look? Also just to flag server performance in terms of rendering the runs can be impacted by the number of hparams we log so we need to be mindful there. In your screenshot we are talking about ~100 variables logged ?

Yes, the number of variables values are added as hparams. Here is an mlflow example run: 
https://mlflow.ecmwf.int/#/experiments/221/runs/6f1b5dd4597742709171a7ed442dd33a

File: training/src/anemoi/training/train/forecaster/forecaster.py (hunk 1 above)
Comment (jakob-schloer) [reply training/src/anemoi/training/train/forecaster/forecaster.py]: @sahahner I went for `variable_loss_scaling.XXX` but dropped the `final_`.

Comment (sahahner) [reply training/src/anemoi/training/train/forecaster/forecaster.py]: I looked at your mlflow example and I think it is sufficiently clear now that these values are the resulting scaler values.

**Review by HCookie — APPROVED** [review]

File: training/src/anemoi/training/train/tasks/base.py [hunk 2]
```diff
@@ -593,3 +596,21 @@ def configure_optimizers(self) -> tuple[list[torch.optim.Optimizer], list[dict]]
         )
 
         return [optimizer], [{"scheduler": scheduler, "interval": "step"}]
+
+    def setup(self, stage: str) -> None:
+        """Lightning hook that is called after model is initialized but before training starts."""
+        # The two ifs should be separate, but are combined due to pre-commit hook
+        if stage == "fit" and self.trainer.is_global_zero:
+            # Log variable scaling on rank 0
```
Comment (jakob-schloer) [inline training/src/anemoi/training/train/tasks/base.py]: I'm not sure, I like the max_variables_to_log. It would probably be better to have a `max_hparam` in the MLflow class, but then we would have to decide which hparams to drop and which not.

File: training/src/anemoi/training/config/diagnostics/evaluation.yaml [hunk 3]
```diff
@@ -57,6 +57,7 @@ log:
     expand_hyperparams: # Which keys in hyperparams to expand
       - config
     http_max_retries: 35
+    max_num_variable_scalings: 200
```
Comment (jakob-schloer) [inline training/src/anemoi/training/config/diagnostics/evaluation.yaml]: If the keyword is not in the config, it would default to 200 using `hasattr(self.config.diagnostics.log.mlflow, "max_num_variable_scalings")`

File: training/src/anemoi/training/train/tasks/base.py [hunk 4]
```diff
@@ -593,3 +596,21 @@ def configure_optimizers(self) -> tuple[list[torch.optim.Optimizer], list[dict]]
...
+            max_variables_to_log = (
+                self.config.diagnostics.log.mlflow.max_num_variable_scalings
+                if hasattr(self.config.diagnostics.log.mlflow, "max_num_variable_scalings")
+                else 200
```
Comment (anaprietonem) [inline training/src/anemoi/training/train/tasks/base.py]: this 200 could be a constant defined for mlflow under the training/diagnostics/mlflow/logger.py as that's just relevant for this

File: training/src/anemoi/training/train/tasks/base.py (hunk 2 above)
Comment (anaprietonem) [reply training/src/anemoi/training/train/tasks/base.py]: There is now a max `max_params_length` that takes care of limiting the length of parameters coming from the provenance/lineage tracking (so all the metadata) and the config entries. We could move the

```
    config_params = OmegaConf.to_container(convert_to_omegaconf(config), resolve=True)

    logger.log_hyperparams(
        config_params,
        expand_keys=config.diagnostics.log.mlflow.expand_hyperparams,
    )

```
 lines in the get_mlflow_logger function to this setup call and then stack together the config and self._scaling_values_log in one list so MAX_PARAMS_LENGTH would cover both. Thoughts?
 
 could still be good that the if statement in the setup also check that mlflow is enabled. This limit just apply to this logger

**Comment by anaprietonem** [comment]
@jakob-schloer could be good if you can update the branch with main! hopefully conflicts should be minor. If you want to discuss any of my comments feel free to drop me a message or we can have a quick call! thanks again for this work!

File: training/src/anemoi/training/train/tasks/base.py (hunk 2 above)
Comment (jakob-schloer) [reply training/src/anemoi/training/train/tasks/base.py]: Very nice idea. I've combined the hyperparameters from the config and the loss_scaling now in the `setup` function and log them there. This way the `max_params_length` is respected and an additional keyword in the config is not needed.

File: training/src/anemoi/training/train/tasks/base.py [hunk 5]
```diff
@@ -761,3 +765,22 @@ def configure_optimizers(self) -> tuple[list[torch.optim.Optimizer], list[dict]]
...
+
+    def setup(self, stage: str) -> None:
+        """Lightning hook that is called after model is initialized but before training starts."""
+        # The conditions should be separate, but are combined due to pre-commit hook
+        if stage == "fit" and self.trainer.is_global_zero and self.logger is not None:
+            # Log hyperparameters on rank 0
+            hyper_params = OmegaConf.to_container(convert_to_omegaconf(self.config), resolve=True)
+            hyper_params.update({"variable_loss_scaling": self._scaling_values_log})
+            # Expand keys for better visibility
+            expand_keys = OmegaConf.select(
```
Comment (anaprietonem) [inline training/src/anemoi/training/train/tasks/base.py:777]: Didn't know you could do this with OmegaConf, just curious was this for readability or  did you encounter any issues if calling?
```
            self.logger.log_hyperparams(
                hyper_params,
                expand_keys=self.config.diagnostics.log.mlflow.expand_hyperparams,
            )
```

**Review by anaprietonem — APPROVED** [review]
LGTM Jakob, left a comment just to clarify. Otherwise good to be merged, thanks!

File: training/src/anemoi/training/train/tasks/base.py (hunk 5 above)
Comment (jakob-schloer) [reply training/src/anemoi/training/train/tasks/base.py:777]: I've only added that logic in case someone is not using an mlflow logger. In that case, the keyword in the config might not exist, and expand_keys would default to `['config]`.
